### PR TITLE
Generic rate limiter

### PR DESCRIPTION
**Pull request by bkeroackdsc**
This restructures the code to allow for multiple, generic rate limiters to be chained. We will use this to implement per-path, per-IP rate limiting. This also cleans up the code a bit in various places.

Functionality should be unchanged.

### CODE REVIEW

**Comment by hankjacobs** [comment]
Oh hey. 

Just ack'ing your review request. I'll try and get to it ASAP.

File: pkg/guardian/guardian_test.go [hunk 1]
```diff
@@ -23,13 +23,22 @@ func newAcceptanceGuardianServer(t *testing.T, logger logrus.FieldLogger) (*Serv
 
 	stop := make(chan struct{})
 	redis := redis.NewClient(&redis.Options{Addr: mr.Addr()})
+	if res := redis.Ping(); res.Err() != nil {
```
Comment (hankjacobs) [inline pkg/guardian/guardian_test.go]: I assume this was done out of necessity? Super curious as to why. IIRC, mr is a mock redis that runs in process so it should never really go down.

File: pkg/guardian/guardian_test.go [hunk 2]
```diff
@@ -103,6 +112,7 @@ func TestBasicFunctionality(t *testing.T) {
 	}()
 	defer func() {
 		close(stop)
+		time.Sleep(50 * time.Millisecond) // give the redisConfStore.RunSync() goroutine time to exit
```
Comment (hankjacobs) [inline pkg/guardian/guardian_test.go]: ehhhhhhh lol. Not a big fan of sleep and pray things happen in time. Not enough to request changes, just wanted to call it out.

Comment (bkeroackdsc) [reply pkg/guardian/guardian_test.go]: I agree. :) The test failures I was getting seemed mysterious (this was months ago so it's a bit foggy to me ATM) and probably due to a race. Which was confirmed by the "fix" of adding a sleep.

File: pkg/guardian/guardian_test.go (hunk 1 above)
Comment (bkeroackdsc) [reply pkg/guardian/guardian_test.go]: IIRC I was getting connection refused errors prior to this change. 🤷‍♂

**Comment by mikeykhalil** [comment]
Committed these changes in the following PR: https://github.com/dollarshaveclub/guardian/pull/32

Closing

**Pull request closed without merging**